File: MIDTERM-LAB-1/LowPerformingProducts.java
import java.io.*;
import java.util.*;

/**
 * LowPerformingProducts.java
 * Detects low-performing video game products from vgchartz-2024.csv.
 *
 * - Reads EVERY row (64 016 data rows, 14 columns each).
 * - Aggregates total_sales per game title across all platforms/regions.
 * - Games with no sales data in ANY row still appear (0.00 M).
 * - Computes dataset average over all unique titles.
 * - Flags titles whose aggregated sales are below the average.
 *
 * Compile : javac DataRecord.java LowPerformingProducts.java
 * Run     : java LowPerformingProducts
 */
public class LowPerformingProducts {

    // ── CSV row parser (handles quoted fields with embedded commas) ──────────
    private static String[] parseCSVLine(String line) {
        List<String> fields = new ArrayList<>();
        StringBuilder sb    = new StringBuilder();
        boolean inQuotes    = false;
        for (int i = 0; i < line.length(); i++) {
            char c = line.charAt(i);
            if (c == '"') {
                inQuotes = !inQuotes;
            } else if (c == ',' && !inQuotes) {
                fields.add(sb.toString());
                sb.setLength(0);
            } else {
                sb.append(c);
            }
        }
        fields.add(sb.toString());
        return fields.toArray(new String[0]);
    }

    // ── Load every row from the CSV ─────────────────────────────────────────
    public static List<DataRecord> loadCSV(String filePath) throws IOException {
        List<DataRecord> records = new ArrayList<>();
        BufferedReader reader    = new BufferedReader(
            new InputStreamReader(new FileInputStream(filePath), "UTF-8"));

        String headerLine = reader.readLine();
        if (headerLine == null) { reader.close(); throw new IOException("CSV file is empty."); }

        // Map header names → column indices (case-insensitive, trimmed)
        String[] headers = parseCSVLine(headerLine);
        Map<String, Integer> colIdx = new LinkedHashMap<>();
        for (int i = 0; i < headers.length; i++) {
            colIdx.put(headers[i].trim().toLowerCase(), i);
        }

        // Required columns
        String[] required = {"img","title","console","genre","publisher","developer",
                             "critic_score","total_sales","na_sales","jp_sales",
                             "pal_sales","other_sales","release_date","last_update"};
        for (String col : required) {
            if (!colIdx.containsKey(col))
                throw new IOException("Missing expected column: " + col);
        }

        int iImg    = colIdx.get("img");
        int iTitle  = colIdx.get("title");
        int iCon    = colIdx.get("console");
        int iGenre  = colIdx.get("genre");
        int iPub    = colIdx.get("publisher");
        int iDev    = colIdx.get("developer");
        int iCS     = colIdx.get("critic_score");
        int iTS     = colIdx.get("total_sales");
        int iNA     = colIdx.get("na_sales");
        int iJP     = colIdx.get("jp_sales");
        int iPAL    = colIdx.get("pal_sales");
        int iOther  = colIdx.get("other_sales");
        int iRel    = colIdx.get("release_date");
        int iUpd    = colIdx.get("last_update");

        String line;
        int lineNum = 1;
        while ((line = reader.readLine()) != null) {
            lineNum++;
            if (line.trim().isEmpty()) continue;
            String[] c = parseCSVLine(line);

            // Safely get a column value (returns "" if index out of range)
            java.util.function.IntFunction<String> col =
                idx -> (idx < c.length) ? c[idx].trim() : "";

            records.add(new DataRecord(
                col.apply(iImg),
                col.apply(iTitle),
                col.apply(iCon),
                col.apply(iGenre),
                col.apply(iPub),
                col.apply(iDev),
                DataRecord.parseDouble(col.apply(iCS)),
                DataRecord.parseDouble(col.apply(iTS)),
                DataRecord.parseDouble(col.apply(iNA)),
                DataRecord.parseDouble(col.apply(iJP)),
                DataRecord.parseDouble(col.apply(iPAL)),
                DataRecord.parseDouble(col.apply(iOther)),
                col.apply(iRel),
                col.apply(iUpd)
            ));
        }
        reader.close();

        if (records.isEmpty())
            throw new IOException("No valid data rows found in CSV.");
        return records;
    }

    // ── Aggregate total_sales per title (sum over all platform/region rows) ──
    public static LinkedHashMap<String, Double> aggregateSales(List<DataRecord> records) {
        // Use insertion-order map; we'll sort by sales descending after
        Map<String, Double> totals = new LinkedHashMap<>();
        for (DataRecord r : records) {
            String key = r.getTitle();
            totals.merge(key, r.getTotalSales(), Double::sum);
        }

        // Sort descending by total sales
        List<Map.Entry<String, Double>> entries = new ArrayList<>(totals.entrySet());
        entries.sort((a, b) -> Double.compare(b.getValue(), a.getValue()));

        LinkedHashMap<String, Double> sorted = new LinkedHashMap<>();
        for (Map.Entry<String, Double> e : entries) sorted.put(e.getKey(), e.getValue());
        return sorted;
    }

    // ── Average over all unique titles ──────────────────────────────────────
    public static double computeAverage(Map<String, Double> totals) {
        return totals.values().stream().mapToDouble(Double::doubleValue).average().orElse(0.0);
    }

    // ── Separator helper ─────────────────────────────────────────────────────
    private static final int W = 78;
    private static void sep(char ch) {
        System.out.println(String.valueOf(ch).repeat(W));
    }

    // ── Entry point ──────────────────────────────────────────────────────────
    public static void main(String[] args) {
        Scanner input = new Scanner(System.in);
        List<DataRecord> records = null;
        File file;

        // ── 1. File-path input loop ─────────────────────────────────────
        while (true) {
            System.out.print("Enter dataset file path: ");
            String path = input.nextLine();
            file = new File(path);

            if (file.exists() && file.isFile()) {
                try {
                    System.out.println("File found. Processing...");
                    records = loadCSV(path);
                    System.out.printf("  [OK] %,d rows loaded from CSV.%n", records.size());
                    break;
                } catch (IOException e) {
                    System.out.println("Invalid file path. Please try again.");
                }
            } else {
                System.out.println("Invalid file path. Please try again.");
            }
        }
        input.close();

        // ── 2. Aggregate ────────────────────────────────────────────────
        System.out.println("Aggregating sales per title...");
        LinkedHashMap<String, Double> totals = aggregateSales(records);
        double average = computeAverage(totals);
        long withSales = totals.values().stream().filter(v -> v > 0).count();

        // ── 3. Header summary ───────────────────────────────────────────
        System.out.println();
        sep('=');
        System.out.println("          VGChartz 2024 — TOTAL SALES PER PRODUCT REPORT");
        sep('=');
        System.out.printf("  Total CSV rows read        : %,d%n", records.size());
        System.out.printf("  Unique game titles         : %,d%n", totals.size());
        System.out.printf("  Titles with sales data     : %,d%n", withSales);
        System.out.printf("  Titles with no sales data  : %,d  (stored as 0.00 M)%n",
                          totals.size() - withSales);
        System.out.printf("  Dataset average (all titles): %.4f M units%n", average);
        sep('=');
        System.out.printf("  %-52s  %12s%n", "GAME TITLE", "TOTAL SALES");
        sep('-');

        // Print ALL unique titles
        int rank = 1;
        for (Map.Entry<String, Double> e : totals.entrySet()) {
            System.out.printf("  %5d. %-52s  %8.2f M%n",
                              rank++, truncate(e.getKey(), 52), e.getValue());
        }
        sep('-');
        System.out.printf("  %-58s  %8.4f M%n", "DATASET AVERAGE", average);
        sep('=');

        // ── 4. Flag low performers ──────────────────────────────────────
        List<Map.Entry<String, Double>> flagged = new ArrayList<>();
        for (Map.Entry<String, Double> e : totals.entrySet()) {
            if (e.getValue() < average) flagged.add(e);
        }

        System.out.println();
        sep('=');
        System.out.println("     ⚠  LOW-PERFORMING PRODUCTS (TOTAL SALES BELOW DATASET AVERAGE)");
        sep('=');
        System.out.printf("  Average threshold : %.4f M units%n", average);
        System.out.printf("  Flagged titles    : %,d of %,d unique games%n",
                          flagged.size(), totals.size());
        sep('-');
        System.out.printf("  %-52s  %12s%n", "GAME TITLE", "TOTAL SALES");
        sep('-');

        int fRank = 1;
        for (Map.Entry<String, Double> e : flagged) {
            System.out.printf("  %5d. %-52s  %8.2f M%n",
                              fRank++, truncate(e.getKey(), 52), e.getValue());
        }
        sep('=');
        System.out.println();
        System.out.println("  Recommendation: Review and consider phasing out flagged titles.");
        System.out.printf("  (%,d titles flagged out of %,d total unique games in dataset)%n",
                          flagged.size(), totals.size());
        sep('=');
        System.out.println();
    }

    // ── Truncate long strings for display ───────────────────────────────────
    private static String truncate(String s, int max) {
        if (s == null) return "";
        return s.length() <= max ? s : s.substring(0, max - 3) + "...";
    }
}

File: MIDTERM-LAB-1/LowPerformingProducts.js
/**
 * LowPerformingProducts.js
 * Detects low-performing video game products from vgchartz-2024.csv.
 *
 * - Reads EVERY row (64,016 data rows, 14 columns each).
 * - Aggregates total_sales per game title across all platforms/region rows.
 * - Games with no sales data in any row still appear (0.00 M).
 * - Computes dataset average over all unique titles.
 * - Flags titles whose aggregated sales fall below the average.
 *
 * Run: node LowPerformingProducts.js
 */

'use strict';

const fs       = require('fs');
const readline = require('readline');

// ════════════════════════════════════════════════════════════════════════════
// DataRecord — stores all 14 columns from one CSV row
// ════════════════════════════════════════════════════════════════════════════
function DataRecord(img, title, console_, genre, publisher, developer,
                    criticScore, totalSales, naSales, jpSales,
                    palSales, otherSales, releaseDate, lastUpdate) {
    this.img         = img;
    this.title       = title;
    this.console     = console_;
    this.genre       = genre;
    this.publisher   = publisher;
    this.developer   = developer;
    this.criticScore = criticScore;   // number (0 if missing)
    this.totalSales  = totalSales;    // number in millions (0 if missing)
    this.naSales     = naSales;
    this.jpSales     = jpSales;
    this.palSales    = palSales;
    this.otherSales  = otherSales;
    this.releaseDate = releaseDate;
    this.lastUpdate  = lastUpdate;
}

// ════════════════════════════════════════════════════════════════════════════
// Utility helpers
// ════════════════════════════════════════════════════════════════════════════

/** Safely parse a string to float; returns 0 for blank / N/A / non-numeric. */
function safeFloat(s) {
    if (!s) return 0;
    const t = s.trim();
    if (t === '' || t.toLowerCase() === 'n/a') return 0;
    const v = parseFloat(t);
    return isNaN(v) ? 0 : v;
}

/**
 * Parse one CSV line correctly, handling quoted fields that may contain commas.
 * Returns an array of field strings.
 */
function parseCSVLine(line) {
    const fields = [];
    let cur = '';
    let inQ = false;
    for (let i = 0; i < line.length; i++) {
        const ch = line[i];
        if (ch === '"') {
            inQ = !inQ;
        } else if (ch === ',' && !inQ) {
            fields.push(cur);
            cur = '';
        } else {
            cur += ch;
        }
    }
    fields.push(cur);
    return fields;
}

/** Truncate a string to max chars for aligned display. */
function trunc(s, max) {
    if (!s) return '';
    return s.length <= max ? s : s.slice(0, max - 3) + '...';
}

/** Right-pad / left-pad helpers. */
const rpad = (s, n) => String(s).padEnd(n);
const lpad = (s, n) => String(s).padStart(n);

/** Format a number as e.g. "19,700.00 M" */
function fmtM(n) {
    return n.toLocaleString('en-US', { minimumFractionDigits: 2, maximumFractionDigits: 2 }) + ' M';
}

const W   = 80;
const SEP = (ch = '=') => console.log(ch.repeat(W));

// ════════════════════════════════════════════════════════════════════════════
// loadCSV — reads and parses the entire file; returns DataRecord[]
// ════════════════════════════════════════════════════════════════════════════
function loadCSV(filePath) {
    const raw     = fs.readFileSync(filePath, 'utf8');
    const lines   = raw.split(/\r?\n/);

    if (lines.length < 2) throw new Error('CSV file is empty or has only a header.');

    // Parse header
    const headers = parseCSVLine(lines[0]).map(h => h.trim().toLowerCase());

    const need = ['img','title','console','genre','publisher','developer',
                  'critic_score','total_sales','na_sales','jp_sales',
                  'pal_sales','other_sales','release_date','last_update'];

    const idx = {};
    for (const col of need) {
        const i = headers.indexOf(col);
        if (i === -1) throw new Error(`Missing expected column: "${col}"\nDetected headers: ${lines[0]}`);
        idx[col] = i;
    }

    const records = [];
    let skipped   = 0;

    for (let i = 1; i < lines.length; i++) {
        const line = lines[i];
        if (!line.trim()) continue;            // skip blank lines

        const c = parseCSVLine(line);
        const g = col => (c[idx[col]] !== undefined ? c[idx[col]].trim() : '');

        const title = g('title');
        if (!title) { skipped++; continue; }  // skip rows with no title at all

        records.push(new DataRecord(
            g('img'),
            title,
            g('console'),
            g('genre'),
            g('publisher'),
            g('developer'),
            safeFloat(g('critic_score')),
            safeFloat(g('total_sales')),
            safeFloat(g('na_sales')),
            safeFloat(g('jp_sales')),
            safeFloat(g('pal_sales')),
            safeFloat(g('other_sales')),
            g('release_date'),
            g('last_update')
        ));
    }

    if (records.length === 0) throw new Error('No valid data rows found in CSV.');
    if (skipped > 0) console.log(`  [Info] ${skipped} blank/title-less line(s) skipped.`);
    return records;
}

// ════════════════════════════════════════════════════════════════════════════
// aggregateSales — sums total_sales per unique game title, sorted desc
// ════════════════════════════════════════════════════════════════════════════
function aggregateSales(records) {
    const totals = {};
    for (const r of records) {
        totals[r.title] = (totals[r.title] || 0) + r.totalSales;
    }

    // Sort by sales descending
    return Object.entries(totals)
        .sort((a, b) => b[1] - a[1]);
        // returns array of [title, totalSales] pairs
}

// ════════════════════════════════════════════════════════════════════════════
// computeAverage — mean of all unique-title totals
// ════════════════════════════════════════════════════════════════════════════
function computeAverage(sorted) {
    const sum = sorted.reduce((acc, [, v]) => acc + v, 0);
    return sum / sorted.length;
}

// ════════════════════════════════════════════════════════════════════════════
// printFullTable — prints every unique title with its total sales
// ════════════════════════════════════════════════════════════════════════════
function printFullTable(sorted, average, totalRows) {
    const withSales  = sorted.filter(([, v]) => v > 0).length;
    const noSales    = sorted.length - withSales;

    SEP('=');
    console.log('        VGChartz 2024 — TOTAL SALES PER PRODUCT REPORT');
    SEP('=');
    console.log(`  Total CSV rows read         : ${totalRows.toLocaleString()}`);
    console.log(`  Unique game titles          : ${sorted.length.toLocaleString()}`);
    console.log(`  Titles with sales data      : ${withSales.toLocaleString()}`);
    console.log(`  Titles with no sales data   : ${noSales.toLocaleString()}  (stored as 0.00 M)`);
    console.log(`  Dataset average (all titles): ${average.toFixed(4)} M units`);
    SEP('=');
    console.log(`  ${rpad('RANK', 6)} ${rpad('GAME TITLE', 52)}  ${'TOTAL SALES'.padStart(14)}`);
    SEP('-');

    let rank = 1;
    for (const [title, sales] of sorted) {
        console.log(`  ${lpad(rank++, 5)}. ${rpad(trunc(title, 52), 52)}  ${lpad(fmtM(sales), 14)}`);
    }

    SEP('-');
    console.log(`  ${rpad('DATASET AVERAGE', 59)}  ${lpad(average.toFixed(4) + ' M', 14)}`);
    SEP('=');
}

// ════════════════════════════════════════════════════════════════════════════
// printFlagged — prints titles whose sales are below average
// ════════════════════════════════════════════════════════════════════════════
function printFlagged(sorted, average) {
    const flagged = sorted.filter(([, v]) => v < average);

    console.log();
    SEP('=');
    console.log('     ⚠  LOW-PERFORMING PRODUCTS (TOTAL SALES BELOW DATASET AVERAGE)');
    SEP('=');
    console.log(`  Average threshold  : ${average.toFixed(4)} M units`);
    console.log(`  Flagged titles     : ${flagged.length.toLocaleString()} of ${sorted.length.toLocaleString()} unique games`);
    SEP('-');
    console.log(`  ${rpad('RANK', 6)} ${rpad('GAME TITLE', 52)}  ${'TOTAL SALES'.padStart(14)}`);
    SEP('-');

    let rank = 1;
    for (const [title, sales] of flagged) {
        console.log(`  ${lpad(rank++, 5)}. ${rpad(trunc(title, 52), 52)}  ${lpad(fmtM(sales), 14)}`);
    }

    SEP('=');
    console.log();
    console.log('  Recommendation: Review and consider phasing out flagged titles.');
    console.log(`  (${flagged.length.toLocaleString()} titles flagged out of ${sorted.length.toLocaleString()} total unique games in dataset)`);
    SEP('=');
    console.log();
}

// ════════════════════════════════════════════════════════════════════════════
// processData — aggregate, print full table, print flagged
// ════════════════════════════════════════════════════════════════════════════
function processData(records) {
    console.log('Aggregating sales per title...');
    const sorted  = aggregateSales(records);
    const average = computeAverage(sorted);

    console.log();
    printFullTable(sorted, average, records.length);
    printFlagged(sorted, average);
}

// ════════════════════════════════════════════════════════════════════════════
// askFilePath — recursive callback loop (required pattern)
// ════════════════════════════════════════════════════════════════════════════
const rl = readline.createInterface({
    input:  process.stdin,
    output: process.stdout
});

function askFilePath() {
    rl.question('Enter dataset file path: ', function(filePath) {
        if (fs.existsSync(filePath)) {
            try {
                console.log('File found. Processing...');
                const records = loadCSV(filePath);
                console.log(`  [OK] ${records.length.toLocaleString()} rows loaded from CSV.`);
                rl.close();
                processData(records);
            } catch (err) {
                console.log('Invalid file path. Try again.');
                askFilePath();
            }
        } else {
            console.log('Invalid file path. Try again.');
            askFilePath();
        }
    });
}

askFilePath();